### PR TITLE
Improve performance dav PROPFIND

### PR DESCRIPTION
**Pull request by icewind1991**
- Allow reusing known folder info when getting directory content
- Reuse known fileinfo to get the filesystem node for each dav node
- Use known filesystem node from above in `SharesPlugin` instead of re-querying

Saves 4 queries for each propfind in my testing

### CODE REVIEW

File: apps/dav/lib/Connector/Sabre/Node.php [hunk 1]
```diff
@@ -403,6 +418,10 @@ public function getFileInfo() {
 		return $this->info;
 	}
 
+	public function getNode() {
```
Comment (come-nc) [inline apps/dav/lib/Connector/Sabre/Node.php:421]: Return type.
Also I think you can extend it in apps/dav/lib/Connector/Sabre/Directory.php to change the return type to Folder with an instance check. That will help with psalm.

File: lib/private/Files/View.php [hunk 2]
```diff
@@ -1431,7 +1431,7 @@ public function getFileInfo($path, $includeMountPoints = true) {
 	 * @param string $mimetype_filter limit returned content to this mimetype or mimepart
 	 * @return FileInfo[]
 	 */
-	public function getDirectoryContent($directory, $mimetype_filter = '') {
+	public function getDirectoryContent($directory, $mimetype_filter = '', \OCP\Files\FileInfo $directoryInfo = null) {
```
Comment (come-nc) [inline lib/private/Files/View.php:1434]: `use OCP\Files\FileInfo;`

File: lib/private/Files/View.php [hunk 3]
```diff
@@ -1448,13 +1448,20 @@ public function getDirectoryContent($directory, $mimetype_filter = '') {
...
+				$data = $this->getCacheEntry($storage, $internalPath, $directory);
+				if (!$data instanceof ICacheEntry || !isset($data['fileid'])) {
+					return [];
+				}
+			} else {
+				$data = $directoryInfo;
+			}
 
-			if (!$data instanceof ICacheEntry || !isset($data['fileid']) || !($data->getPermissions() && Constants::PERMISSION_READ)) {
+			if (!($data->getPermissions() & Constants::PERMISSION_READ)) {
```
Comment (come-nc) [inline lib/private/Files/View.php:1460]: rebase once https://github.com/nextcloud/server/pull/31754 is merged

**Review by come-nc — APPROVED** [review]

File: apps/dav/lib/Connector/Sabre/Node.php [hunk 4]
```diff
@@ -75,6 +78,8 @@ abstract class Node implements \Sabre\DAV\INode {
 	 */
 	protected $shareManager;
 
+	protected \OC\Files\Node\Node $node;
```
Comment (skjnldsv) [inline apps/dav/lib/Connector/Sabre/Node.php:81]: use \OC\Files\Node\Node

File: lib/private/Files/View.php (hunk 2 above)
Comment (icewind1991) [reply lib/private/Files/View.php:1434]: That would conflict with `\OC\Files\FileInfo`

File: apps/dav/lib/Connector/Sabre/Node.php (hunk 4 above)
Comment (icewind1991) [reply apps/dav/lib/Connector/Sabre/Node.php:81]: Would conflict with the `Node` class here

**Comment by skjnldsv** [comment]
Needs a rebase and we're good! :rocket:

**Review by skjnldsv — APPROVED** [review]

**Pull request merged**